server/src/main/java/org/cloudfoundry/identity/uaa/approval/VmidentityApprovalStore.java

```diff
@@ -13,138 +13,140 @@
+import org.cloudfoundry.identity.uaa.client.VmidentityDataAccessException;
+import org.cloudfoundry.identity.uaa.util.VmidentityUtils;
 import org.springframework.context.ApplicationEvent;
 import org.springframework.context.ApplicationEventPublisher;
 import org.springframework.context.ApplicationEventPublisherAware;
 import org.springframework.security.core.Authentication;
 import org.springframework.security.core.context.SecurityContextHolder;
 
 public class VmidentityApprovalStore implements ApprovalStore, ApplicationEventPublisherAware {
-
-    // in memory for now
-    private List<Approval> _approvals;
-
     private final Log logger = LogFactory.getLog(VmidentityApprovalStore.class);
 
-    // USER_APPROVALS_FILTER_TEMPLATE = "user_id eq \"%s\""
-    // USER_FILTER_TEMPLATE = "user_id eq \"%s\"";
-    // USER_AND_CLIENT_FILTER_TEMPLATE = "user_id eq \"%s\" and client_id eq \"%s\"";
-
-    private final Object _lock;
+    private ApplicationEventPublisher applicationEventPublisher;
+    private CasIdmClient client;
+    private boolean handleRevocationsAsExpiry = false;
 
-    private ApplicationEventPublisher _applicationEventPublisher;
+    public VmidentityApprovalStore(CasIdmClient client) {
+        this.client = client;
+    }
 
-    public VmidentityApprovalStore() {
-        this._approvals = new ArrayList<Approval>();
-        this._lock = new Object();
+    public void setHandleRevocationsAsExpiry(boolean handleRevocationsAsExpiry) {
+        this.handleRevocationsAsExpiry = handleRevocationsAsExpiry;
     }
 
     @Override
     public boolean addApproval(Approval approval) {
-        synchronized (this._lock) {
-            this._approvals.add(approval);
+        try {
+            String tenant = VmidentityUtils.getTenantName(client);
+            Approval added = convertToUaaApproval(client.addApproval(tenant, convertToIdmApproval(approval)));
+
+            if (added != null) {
+                Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+                publish(new ApprovalModifiedEvent(added, authentication));
+                return true;
+            } else {
+                return false;
+            }
+        } catch (Exception ex) {
+            logger.error("Adding an approval failed", ex);
+            throw new VmidentityDataAccessException("Unable to add approval");
         }
-        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
-        publish(new ApprovalModifiedEvent(approval, authentication));
-        return true;
     }
 
     @Override
     public boolean revokeApproval(Approval approval) {
-        synchronized (this._lock) {
-            this._approvals.remove(approval);
-        }
-        return false;
+        return revokeApprovals(String.format("user_id eq \"%s\" and client_id eq \"%s\" and scope eq \"%s\"", approval.getUserId(), approval.getClientId(), approval.getScope()));
```

Comment:
this seems a bit of a backwards direction, if we have approval it seems much more efficient to operate over it, rather than create a filter,m and then go parse it etc. ... ?